### ISSUE TITLE
Climate sensor becomes unavailable

### ISSUE THREAD

**Issue by argoyle** [sfalkman/ngenic-hass-platform#25]
Hi!

Installed the integration via HACS and all seemed fined. After a quite short while the climate sensor became unavilable. Restarted HA and the sensor was fine again but a short while later (less than an hour) the sensor became unavailable again.

The other sensors seems fine.

The logs shows a connection reset by peer.

Logger: custom_components.ngenic.climate
Source: custom_components/ngenic/climate.py:144
Integration: Ngenic Tune (documentation)
First occurred: 10:02:27 (72 occurrences)
Last logged: 15:57:27

Failed to update climate '4dbec961-ec30-eb11-9fb4-000d3ab212db-climate'
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/httpx/_exceptions.py", line 326, in map_exceptions
    yield
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1502, in _send_single_request
    (status_code, headers, stream, ext,) = await transport.arequest(
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/connection_pool.py", line 199, in arequest
    await self._keepalive_sweep()
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/connection_pool.py", line 339, in _keepalive_sweep
    await connection.aclose()
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/connection.py", line 183, in aclose
    await self.connection.aclose()
  File "/usr/local/lib/python3.8/site-packages/httpcore/_async/http11.py", line 202, in aclose
    await self.socket.aclose()
  File "/usr/local/lib/python3.8/site-packages/httpcore/_backends/asyncio.py", line 198, in aclose
    await self.stream_writer.wait_closed()  # type: ignore
  File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.8/site-packages/httpcore/_exceptions.py", line 12, in map_exceptions
    raise to_exc(exc) from None
httpcore.CloseError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/ngenicpy/models/base.py", line 172, in _async_request
    r = await request_method(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1548, in get
    return await self.request(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1371, in request
    response = await self.send(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1406, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1444, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1476, in _send_handling_redirects
    response = await self._send_single_request(request, timeout)
  File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1502, in _send_single_request
    (status_code, headers, stream, ext,) = await transport.arequest(
  File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.8/site-packages/httpx/_exceptions.py", line 343, in map_exceptions
    raise mapped_exc(message, **kwargs) from exc  # type: ignore
httpx.CloseError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/ngenic/climate.py", line 144, in _async_update
    current = await self._node.async_measurement(MeasurementType.TEMPERATURE)
  File "/usr/local/lib/python3.8/site-packages/ngenicpy/models/node.py", line 160, in async_measurement
    return await self._async_parse_new_instance(url, Measurement, node=self, measurement_type=measurement_type)
  File "/usr/local/lib/python3.8/site-packages/ngenicpy/models/base.py", line 120, in _async_parse_new_instance
    ret_json = self._parse(await self._async_get(url))
  File "/usr/local/lib/python3.8/site-packages/ngenicpy/models/base.py", line 238, in _async_get
    return await self._async_request("get",
  File "/usr/local/lib/python3.8/site-packages/ngenicpy/models/base.py", line 179, in _async_request
    raise ClientException(self._get_error("A request exception occurred", r, parent_ex=exc))
ngenicpy.exceptions.ClientException: (ClientException(...), 'A request exception occurred: [Errno 104] Connection reset by peer')

**Comment by sfalkman** [comment]
Hmm, that's strange. Unfortunately I've had a corrupt homeassistant db so I didn't have any history on my own climate, to see if I've had the same issue (didn't find anything in the logs either).

When it gets in the unavailable state, what happens if you wait 5 minutes? It should try to update the climate sensor again -- do you get the same error again, or does the sensor become available again?

**Comment by argoyle** [comment]
It has happened 130 times in the last ~11 hours. See attached screenshot.
![image](https://user-images.githubusercontent.com/328334/107563720-b4cbc980-6be1-11eb-9e44-fdf7d9115171.png)

**Comment by niklaskb** [comment]
I can confirm, I have the same issue. The temperature sensor updates for some time after a full restart, then the updates stop and gives the above error. The humidity sensor seems to be unaffected.

**Comment by sfalkman** [comment]
That is strange, I've been holding back a bit to see if I get the same issue, but I'm not.. I have a couple of timeout errors, but the sensor will start updating again the next iteration.

Maybe httpx can't recover after a "Connection reset by peer" error, I need to find a way to reproduce it..

**Comment by sfalkman** [comment]
What version of homeassistant are you running?

**Comment by sfalkman** [comment]
I will run the below script over night and see if I get any issues. If you'd like to help out you can do it too.  
Make sure to install ngenicpy: `pip3 install ngenicpy` and verify your httpx version: `pip3 show httpx`, it should be `0.16.1`.

```python
from time import gmtime, strftime, sleep
import asyncio
from ngenicpy import AsyncNgenic
from ngenicpy.models.measurement import MeasurementType

async def run():
    async with AsyncNgenic(token="YOUR_TOKEN") as ngenic:
        tunes = await ngenic.async_tunes()
        tune = await ngenic.async_tune(tunes[0].uuid())
        
        control_room = await tune.async_room(tune["roomToControlUuid"])
        control_node = await tune.async_node(control_room["nodeUuid"])

        while True:
          current = await control_node.async_measurement(MeasurementType.TEMPERATURE)
          print("%s %s" % (strftime("%Y-%m-%d %H:%M:%S", gmtime()), current.json()))
          sleep(60)

asyncio.run(run())
```

**Comment by niklaskb** [comment]
Currently 2021.2.3

From the history it looks like i’ve had this issue since February 6. I updated to 2021.2.1 around then.

**Comment by sfalkman** [comment]
> Currently 2021.2.3
> 
> From the history it looks like i’ve had this issue since February 6. I updated to 2021.2.1 around then.

Was it after you upgraded the ngenic integration it started happening?

**Comment by niklaskb** [comment]
Not sure, I didn’t discover the issue until recently.

**Comment by argoyle** [comment]
Also running 2021.2.3 and I have had this problem since first install a week ago. It seems to just run fine the first 5 minutes and then crashes. I've started the script and will add the log here tomorrow morning.

**Comment by argoyle** [comment]
It has been running fine all night. Just realized that I ran it over night on my Mac rather than on the NUC where I run HA though (😜 ) so restarted it there now.

**Comment by niklas-izettle** [comment]
Same here. It ran successfully over night on my mac.

**Comment by niklaskb** [comment]
Same here. It ran successfully over night on my mac.

**Comment by sfalkman** [comment]
Yup, it ran fine for me as well.  

Just to recap
* You both are running HomeAssistant 2021.2.3
* You both are getting the error "[Errno 104] Connection reset by peer"

Some questions for you both that might narrow this down:
1. Do you have an Ngenic Track?
2. How are you running hass? Are you running HassOS? Docker?
3. Are you running any other applications that are utilizing the Ngenic API (except for the Ngenic app)?

**Comment by argoyle** [comment]
1. Nope
2. Supervised install on Ubuntu (which is a dockerized install)
3. Nope

Script has run fine for 45 minutes on the NUC now as well. Will try to run it "inside" the core-container as well now. Feels like it is something that HA affects.

**Comment by argoyle** [comment]
Still find it strange that it's only the climate-sensor that has this behaviour though.

**Comment by sfalkman** [comment]
I managed to reproduce the issue.  
It works for me in hass because I have a track, which has a sensor that updates each minute -- this will keep the connection to the Ngenic server alive. If you don't have a track, the climate will update every 5 minutes but the server will have closed the connection by then (causing Connection reset by peer). This will probably purge the connection from the httpx pool, making subsequent requests (from regular sensors) to initiate a new connection to the server, which will succeed.

I changed the script from above to `sleep(400)` and got the same issue.

I'll look into this as soon as I can, hopefully tonight.

**Comment by argoyle** [comment]
Oooh, nice catch. Looking forward to being able to test it. No rush though. 😊

**Comment by sfalkman** [comment]
Just an update.. I've tried working around this with httpx, but there seems to be a bug in httpx.

My _guess_ is that the ngenic server somehow destroys the connection without gracefully closing it, and this doesn't play well with httpx's asynchronous client. When a request is sent with httpx it first cleans up connections that have timed out (according to httpx's timeout, which is 5 seconds). Httpx tries to do a graceful `close()` on these connections, but if a connection have been destroyed by the ngenic server, the `close()` seems to crash.  
I've tried to reproduce it myself, but it's hard to produce such error where the connection is destroyed (causing a "Connection reset by peer"). If I find a way, I will report it on the httpx project.

To work around this, I could probably add a retry if this specific error occurs. It's a stupid solution, but a solution nonetheless.

**Comment by argoyle** [comment]
Would it help to just run it more often than every 5 minutes? It seemed to run fine every minute even if it only changed value every 5 minutes in the script.

**Comment by argoyle** [comment]
Also a stupid solution. 😁

**Comment by sfalkman** [comment]
Yes, that would help, but I'm also maintaining ngenicpy that the homeassistant component is using, and ngenicpy has the same issue with broken connections so I rather "fix" it there :)

I've pushed a fix, could you try updating the component and see if it works?

**Comment by argoyle** [comment]
Fully understandable. Running now. Will report back shortly.

**Comment by argoyle** [comment]
Has been running fine for two hours now. 😊 Thanks for fixing.

**Comment by sfalkman** [comment]
Good to hear! I'll leave the bug open over the weekend, then I'll close it if you've not reported any issues.

**Comment by niklaskb** [comment]
I can also confirm that it's fixed, It has been running the updated version 27 hours now without issues.

**Comment by sfalkman** [comment]
Fixed with ngenicpy 0.3.3
